shared/src/utils/random.ts

```diff
@@ -97,14 +164,31 @@ export default class Random {
 
   /**
    * Chooses a random element from an array of weighted elements.
-   *
-   * @param array - An array of tuples of the form [element, weight]
+   * @param elements
+   * @param weights
    * @returns A random element from the array.
    */
-  weightedChoice<T>(array: Array<[element: T, weight: number]>): T {
-    const elements = array.map((w) => w[0])
-    const weigths = array.map((w) => w[1])
-    return elements[this.weightedIndex(weigths)]
+  weightedChoice<T>(elements: ReadonlyArray<T>, weights: number[]): T
+  /**
+   * Chooses a random element from an array of weighted elements.
+   * @param choicesAndWeights An array of element-weight tuples
+   * @returns A random element from the array.
+   */
+  weightedChoice<T>(choicesAndWeights: ReadonlyArray<[element: T, weight: number]>): T
+  weightedChoice<T>(
+    data: ReadonlyArray<[element: T, weight: number]> | ReadonlyArray<T>,
+    weights: number[] | undefined = undefined,
+  ): T {
+    let elements: readonly T[]
+
+    if (weights === undefined) {
+      elements = (data as ReadonlyArray<[element: T, weight: number]>).map((w) => w[0])
```

Comment:
Since `ReadonlyArray<>` is a type in typescript, there seems to be no way to check which type we actually received. This seems only to be possible for the primitive types, where `typeof` works, or for classes where `instanceof` works. For objects one has to rely on some unique properties that could be used to differentiate between object types. Neither works in this case. However, typescript already guaranties that the function is called either as `Readonly<[el, w]>, undefined` or as `Readonly<T>, number[]`. So this is where the type checking already happens. So I used `weights` as an indicator to decide which type data has.

Maybe a custom type guard like the following in more in the typescript spirit, than using `as`

```
function isSingleValueCall(_: any, weights: any): _ is ReadonlyArray<[element: T, weight: number]> {
  return weights === undefined
}
```
using this function typescript can determine the right type. But it is still the same check as before.

I changed it to use the function. I think it is more readable and shows the intend of the code. I also added the checks for empty data and a check if each value has a weight and throw an error if necessary.